SWC-3836 - Invite a new user to join a team (Phase 1)

hhu94 opened the pull request:
- [x] Allow users to send email invitations to new users.
- [x] Handle invitations links sent out to new user.
  - [x] Signing in
  - [x] Registering a new account
- [x] Tests

Commit:
Add ability to send email invitation to new user

NOTE: this commit contains a lot of code duplication

Commit:
Undo pom.xml synapse.version change

Commit:
Undo Portal.gwt.xml change

Commit:
Display email invitations in team page

Commit:
Tweak appearances

Commit:
Show invitee email even when invitee id is set

Commit:
Revert "Add ability to send email invitation to new user"

This reverts commit 8b61a57b5da036a69ff2449e37b70cd8408c927c.

Commit:
Enable sending email invitations to new users

Rewriting of commit 8b61a57b5da036a69ff2449e37b70cd8408c927c.

jay-hodgson commented:
Refactoring looks good so far.  But why do you go through the Suggestion provider when the user enters an email address?  Does this approach mean that the user needs to select the email address in the dropdown suggestion list after they paste it into the textbox?  
Alternatively, could you not look at the textbox text to see if it matches your email regex.  And if so, move forward with the invitation using an email (rather than depending on the suggestion provider callback, which now includes the email from your changes?).

Commit:
Change InviteWidget behavior

No longer need to select suggestion when entering an email address,
nor when entering a username that exactly matches a Synapse user's.

Add tests for these behaviors.

Commit:
Refactor RegisterWidget

Commit:
Add prototype impl of invitation handling

Commit:
Reset selected suggestion on input change

Commit:
Flesh out the rest of the feature

This commit is dependent on https://github.com/hhu94/Synapse-Repository-Services/commit/fb91569f31fd74179cdf6fd8f4c8f2406dfa56b9

Note that synapse.version is not yet updated in the pom.xml, I will do
so when the above PLFM commit is merged into develop

Commit:
Merge branch 'develop' into SWC-3836

jay-hodgson commented on pom.xml:
```diff
@@ -633,7 +633,13 @@
 		</dependency>
-	</dependencies>
+		<dependency>
+			<groupId>com.googlecode.gwtquery</groupId>
```
How much does this add to the client download (the gwt wrapper for jquery)?  If we can incorporate (to avoid "callback hell" by using promises), maybe it's worth it.

jay-hodgson replied:
Is it well maintained, does it lock us into a version of jquery that it wraps?

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/PortalGinInjector.java:
```diff
@@ -243,315 +93,506 @@
-	public ProfilePresenter getProfilePresenter();
+	GlobalApplicationState getGlobalApplicationState();
+
+	PresenterProxy<HomePresenter, Home> getHomePresenter();
```
Why all the extra newlines?  Like the removal of implicit "public" in interface, but the large amount of cleanup-related changes should probably be in a separate PR.

hhu94 replied:
I undid the new lines and kept the removal of the public keywords.

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/PortalGinModule.java:
```diff
@@ -163,342 +46,48 @@
-import org.sagebionetworks.web.client.widget.discussion.SingleDiscussionThreadWidgetView;
-import org.sagebionetworks.web.client.widget.discussion.SingleDiscussionThreadWidgetViewImpl;
-import org.sagebionetworks.web.client.widget.discussion.SubscribersWidgetView;
-import org.sagebionetworks.web.client.widget.discussion.SubscribersWidgetViewImpl;
```
Again, nice cleanup, but as soon as anyone runs "Organize Imports" within the IDE all of these will show up again.

hhu94 replied:
I fixed all my imports to match that of Eclipse's auto importing behavior.

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/SynapseClient.java:
```diff
@@ -372,7 +375,9 @@ EntityBundlePlus getEntityBundlePlusForVersion(String entityId, Long versionNumb
 	Entity moveEntity(String entityId, String newParentEntityId) throws RestServiceException;
 
-	String getUserIdFromUsername(String username) throws RestServiceException;
+    Team getTeam(String teamId) throws RestServiceException;
```
Whitespace is off for these (if these are being cleaned up, then this should also be right).

hhu94 replied:
I fixed all my files to be indented with tabs. (Sometimes using spaces for alignment). I wish there was a way to show github diffs with tabsize = 4 instead of 8, because that's what I use in my IDE (Intellij IDEA).

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java:
```diff
@@ -0,0 +1,158 @@
+		this.authController = authController;
+		this.placeChanger = globalApplicationState.getPlaceChanger();
+		view.setPresenter(this);
+		registerWidget.enableEmailAddressField(false);
```
Initialization of view state should not be in the constructor (since state will be lost when it's reused).  Please move to init (in setPlace)

hhu94 replied:
I changed how the view is initialized, but I'd like to discuss what I did further in person.

jay-hodgson commented on src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java:
```diff
@@ -1609,24 +1538,72 @@ public Long getOpenRequestCount(String currentUserId, String teamId)
-	
+
+	@Override
+	public MembershipInvtnSubmission getMembershipInvitation(MembershipInvtnSignedToken token) throws RestServiceException {
```
Simple new methods like this should be done in the SynapseJavascriptClient instead of introducing new GWT RPC calls.

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java:
```diff
@@ -0,0 +1,158 @@
+	}
+
+	private void bindInvitationToAuthenticatedUser(final String misId) {
+		PromiseRPC<InviteeVerificationSignedToken> promise = new PromiseRPC<>();
```
Please format new code, indentation should be done in either tabs and spaces (I think default for Eclipse Format or Correct Indentation uses tabs).

hhu94 replied:
This should be fixed

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/presenter/NewAccountPresenter.java:
```diff
@@ -138,6 +131,9 @@ public void onSuccess(String sessionToken) {
 				//success, send to login place to continue login process (sign terms of use...)
 				view.showInfo(DisplayConstants.ACCOUNT_CREATED, "");
+				if (accountCreationToken.getEncodedMembershipInvtnSignedToken() != null) {
+					globalAppState.setLastPlace(new EmailInvitation(accountCreationToken.getEncodedMembershipInvtnSignedToken()));
```
nifty

hhu94 replied:
This actually took me a bit to figure out haha.

Commit:
Fix whitespace, imports and view state

Commit:
Fix imports and indentation

Commit:
Even more import fixing

Commit:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/SynapseWebClient into SWC-3836

Commit:
Remove gwtquery dependency

Commit:
Move getMembershipInvitation to jsClient

Commit:
Upgrade to GWT 2.8.1, Java 1.8 and Guava 23.0

Commit:
Use Guava FluentFuture to avoid callback hell

Commit:
Add comments explaining upgrade changes

Commit:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/SynapseWebClient into SWC-3836

Commit:
Reimplement EmailInvitationPresenter with Futures

Commit:
Move services from futureClient to jsClient

Commit:
Merge remote-tracking branch 'upstream/develop' into SWC-3836

Commit:
Remove unused getFailedFuture

Commit:
Add tests for EmailInvitationPresenter

Commit:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/SynapseWebClient into SWC-3836

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/PortalGinInjector.java:
```diff
@@ -251,240 +251,327 @@
+	SubjectWidget getSubjectWidget();
+	TeamMemberRowWidget getTeamMemberRowWidget();
+	RequestBuilderWrapper getRequestBuilder();
+=======
```
Doesn't this fail to compile?  (these are usually introduced during a complex merge)

hhu94 replied:
Yes, I didn't make the merge correctly, fixing it up now.

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java:
```diff
@@ -0,0 +1,169 @@
+
+	@Inject
+	public EmailInvitationPresenter(EmailInvitationView view,
+									RegisterWidget registerWidget,
```
all this cleanup, and we get 9 tabs indentation here? :)

hhu94 replied:
This looks aligned when viewed with 4 space tabs. 
![2017-10-26-11 35 55](https://user-images.githubusercontent.com/7145349/32070682-fb2251ec-ba41-11e7-8691-a0169b22fd36.png)
Should I avoid this kind of indentation?

jay-hodgson replied:
I'd recommend having the first parameter on a newline, and just indent all of these dependencies with a single tab.  But it's trivial, so feel free to ignore :)

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java:
```diff
@@ -0,0 +1,169 @@
+		registerWidget.enableEmailAddressField(false);
+		view.setRegisterWidget(registerWidget.asWidget());
+		view.setSynapseAlertContainer(synapseAlert.asWidget());
+		EmailInvitationPresenter.this.registerWidget.setEncodedMembershipInvtnSignedToken(encodedMISignedToken);
```
"EmailInvitationPresenter.this." necessary?

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java:
```diff
@@ -0,0 +1,169 @@
+	private void bindInvitationToAuthenticatedUser(final String misId) {
+		jsClient.getInviteeVerificationSignedToken(misId)
+			.transformAsync(
+					jsClient::updateInviteeId,
```
I need your help understanding this.
"jsClient::updateInviteeId"
Could you explain this syntax?
Calling getInviteeVerificationSignedToken with the MembershipInvtnSubmission ID binds it to the currently authenticated user?

hhu94 replied:
This is passing updateInviteeId as a method reference to transformAsync. It's equivalent to the lambda "token -> jsClient.updateInviteeId(token)". Since the first parameter of transformAsync (AsyncFunction) is a functional interface, it can take in a method reference with one parameter (like above), a lambda, or an instance of an AsyncFunction class.

jay-hodgson replied:
Please change to lambda

Commit:
Revert "Revert "SWC-3644 - Upgrade to Java 8, GWT 2.8.1 and Guava 23.0""

This reverts commit 37d7a6841b58be633c4d9fc94725dde8eeb82f24.

Commit:
Fix bad merge

Commit:
Fix style

jay-hodgson commented on src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java:
```diff
@@ -1610,24 +1627,62 @@ public Long getOpenRequestCount(String currentUserId, String teamId)
-	
+
+	@Override
+	public InviteeVerificationSignedToken getInviteeVerificationSignedToken(String membershipInvitationId) throws RestServiceException {
```
please remove if in js client

jay-hodgson commented on src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java:
```diff
@@ -1610,24 +1627,62 @@ public Long getOpenRequestCount(String currentUserId, String teamId)
+
+	@Override
+	public void updateInviteeId(String membershipInvitationId, InviteeVerificationSignedToken token) throws RestServiceException {
+		org.sagebionetworks.client.SynapseClient synapseClient = createSynapseClient();
```
please remove if in js client

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/widget/team/InviteWidget.java:
```diff
@@ -24,13 +25,14 @@
 	private SynapseAlert synAlert;
 	private SynapseSuggestBox peopleSuggestWidget;
-	
+	private static RegExp emailRegExp = RegExp.compile("[a-z0-9._%+-]+@[a-z0-9.-]+\\.[a-z]{2,4}");
```
Use WebConstants.VALID_EMAIL_REGEX

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java:
```diff
@@ -0,0 +1,170 @@
+	}
+
+	private void bindInvitationToAuthenticatedUser(final String misId) {
+		jsClient.getInviteeVerificationSignedToken(misId)
```
show loading during this process

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java:
```diff
@@ -0,0 +1,170 @@
+	}
+
+	private void initializeView(final MembershipInvtnSubmission mis) {
+		view.show();
```
rename to showNotLoggedInUI()

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java:
```diff
@@ -0,0 +1,170 @@
+			AuthenticationController authController,
+			GlobalApplicationState globalApplicationState) {
+		this.view = view;
+		this.registerWidget = registerWidget;
```
initialize registerWidget.  
enableEmailAddressField(false), add to view.  Add synapse alert to view.

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/view/users/RegisterWidgetViewImpl.java:
```diff
@@ -20,40 +21,30 @@
-			@Override
-			public void onClick(ClickEvent event) {
-				presenter.registerUser(emailAddressField2.getValue());
+				NewUser newUser = new NewUser();
```
callback to the presenter with emailAddressField value.  Presenter will get encoded membership invtn signed token, and construct NewUser object.

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/view/users/RegisterWidgetViewImpl.java:
```diff
@@ -20,40 +21,30 @@
-			public void onClick(ClickEvent event) {
-				presenter.registerUser(emailAddressField2.getValue());
+				NewUser newUser = new NewUser();
+				newUser.setEmail(emailAddressField.getValue());
```
Presenter should probably validate email address before attempting to register.

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/presenter/NewAccountPresenter.java:
```diff
@@ -138,6 +136,9 @@ public void onSuccess(String sessionToken) {
 				view.setLoading(false);
```
Remove place.toToken().contains("email") handling (around line 76)

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/SynapseJavascriptFactory.java:
```diff
@@ -199,6 +203,10 @@ public Object newInstance(OBJECT_TYPE type, JSONObjectAdapter json) throws JSONO
 			return list;
+		case MembershipInvtnSubmission:
+			return new MembershipInvtnSubmission(json);
+			case InviteeVerificationSignedToken:
```
check whitespace

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidgetViewImpl.java:
```diff
@@ -70,22 +70,35 @@ public void configure(List<UserProfile> profiles, List<MembershipInvtnSubmission
+					td.add(inviteeEmailDiv);
+				}
+			} else if (inviteeEmail != null) {
+				userBadge.configureWithInviteeEmail(invite.getInviteeEmail());
```
Consider emailBadge for this

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidgetViewImpl.java:
```diff
@@ -70,22 +70,35 @@ public void configure(List<UserProfile> profiles, List<MembershipInvtnSubmission
+			final MembershipInvtnSubmission invite = invitations.get(i);
+			String inviteeEmail = invite.getInviteeEmail();
+			UserBadge userBadge = ginInjector.getUserBadgeWidget();
+			if (profile != null) {
```
Logic should be moved to presenter.  This can be deferred for this PR.

jay-hodgson commented on src/main/resources/org/sagebionetworks/web/client/view/EmailInvitationViewImpl.ui.xml:
```diff
@@ -0,0 +1,37 @@
+				<g:SimplePanel ui:field="synapseAlertContainer"/>
+			</b:Column>
+		</b:Row>
+		<div class="container margin-top-15 min-height-400">
```
Remove HTMLPanel, remove div, add class names to Container below instead (addStyleName="margin-top-15 min-height-400")

jay-hodgson commented on src/main/resources/org/sagebionetworks/web/client/view/EmailInvitationViewImpl.ui.xml:
```diff
@@ -0,0 +1,37 @@
+						<b:BlockQuote ui:field="invitationMessageWrapper" visible="false">
+							<b:Heading ui:field="invitationMessage" size="H5" addStyleNames="margin-top-10"/>
+						</b:BlockQuote>
+						<b:Row>
```
remove Rows

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/utils/FutureUtils.java:
```diff
@@ -0,0 +1,57 @@
+	 *
+	 * @return a failed future with its exception set to a new Throwable
+	 */
+	public static <T> FluentFuture<T> getFailedFuture() {
```
I wonder if this should take the exception (optionally), then we can test using different error states.

jay-hodgson commented on src/test/java/org/sagebionetworks/web/unitclient/presenter/EmailInvitationPresenterTest.java:
```diff
@@ -0,0 +1,133 @@
+		when(jsClient.getInviteeVerificationSignedToken(mis.getId())).thenReturn(getDoneFuture(inviteeVerificationSignedToken));
+		when(jsClient.updateInviteeId(inviteeVerificationSignedToken)).thenReturn(getDoneFuture(null));
+		presenter.setPlace(place);
+		verify(placeChanger).goTo(any(Profile.class));
```
verify jsClient methods are called.  could capture Profile place as well, and verify it goes to the Teams area.

jay-hodgson commented on src/test/java/org/sagebionetworks/web/unitclient/widget/entity/team/InviteWidgetTest.java:
```diff
@@ -93,8 +93,21 @@ public void testSendToMember() {
+
+	@Test
+	public void testSendToEmailAddress() {
+		String email = "test@example.com";
```
mix case :)

Commit:
Use existing isValidEmail

Commit:
Remove unused services

Commit:
Replace method reference with lambda for clarity

Commit:
Refactor presenter initialization logic

Commit:
Clear view before setting presenter

Commit:
Remove unused NewAccount code

Commit:
Fix indentation

Commit:
Allow passing in a Throwable to getFailedFuture

Commit:
Verify method calls more explictly in test

Commit:
Add showLoading and hideLoading, refactor UI code

Commit:
Move logic from view to presenter

Commit:
Fix tests

Commit:
Undo accidental change

Commit:
Make container non-fluid

Commit:
Add EmailInvitationBadge

Commit:
Refactor OpenUserInvitationsWidget

Commit:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/SynapseWebClient into SWC-3836

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidget.java:
```diff
@@ -79,47 +81,67 @@ public void onFailure(Throwable caught) {
 
 	@Override
 	public void getNextBatch() {
 		synAlert.clear();
```
This interaction confuses me.  I would expect that you would need to hide the More button before the rpc.

hhu94 replied:
I agree that we should hide the more button before the rpc, but which interaction is confusing other than that?

hhu94 replied:
Is it the synAlert.clear()? I'm not entirely sure why it was there before the refactor either. I think it's safe to remove it.

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidget.java:
```diff
@@ -79,47 +81,67 @@ public void onFailure(Throwable caught) {
-	
+
+	private void addInvitations(List<OpenTeamInvitationBundle> bundles) {
+		view.hideInvitations();
```
Does this mean that all invitations hidden while these new invitation widgets are created, and then they are all shown again (if so, why?)?

hhu94 replied:
You're right, this hiding + showing is unnecessary.

Commit:
Remove unnecessary code, hide moreButton on click

Commit:
Fix tests

Commit:
Undo debug mode (again)
